src/Common/src/TypeSystem/Common/TypeSystemHelpers.cs

```diff
@@ -36,18 +36,24 @@ static public PointerType MakePointerType(this TypeDesc type)
 
         static public DefType GetClosestDefType(this TypeDesc type)
         {
-            if (type is DefType)
-                return (DefType)type;
-            else
-                return type.BaseType;
+            // The implementation for this should mirror GetClosestMetadataType
+            throw new NotImplementedException();
         }
 
         static public MetadataType GetClosestMetadataType(this TypeDesc type)
         {
-            if (type is MetadataType)
-                return (MetadataType)type;
-            else
-                return type.BaseType.GetClosestMetadataType();
+            if (type.IsSzArray)
+            {
+                MetadataType arrayType = type.Context.SystemModule.GetType("System", "Array`1");
```

Comment:
`TypeDesc.GetMethods()` returns information that the type owns. The type does not need to consult any other types to give you the answer. `TypeDesc.RuntimeInterfaces` returns derived information that is computed by plugged-in algorithm from the type itself, from other types, and from type system supplied policy. The virtual method resolution is derived information like `TypeDesc.RuntimeInterfaces`. That's why it is also a good candidate for plugged-in algorithm.

From a different angle: I think about this as ICastable that we happen to do at compile time. If we choose to trade perf for size (IMHO, it would be a good trade-off), we should be able to supply the built-in generic array interfaces by implementing ICastable on System.Array and manufacturing the generic interface implementations at runtime using ICastable. It is ok for such change to affect the EEType building code or the algorithms plugged into type system by the compiler. But it should not affect the common type system code - if the common type system is factored properly.


> We would still need magic special casing (GetClosestMetadataType again?) when building vtables.

I think that it is fine to special case the vtable building code for array to handle this for now. If we get more instance of the compile-time ICastables in future, we can generalize it (it is hard to get the generalization right today because of we have just one case). Or we choose to prefer the above trade off and it will go away too.


> And the logic around GetVirtualMethodSlot will be off because the owning type is a weird type

I believe we should not be hitting this case today. I would just add assert or throwing precondition for it for now.